### PR TITLE
Keep alias lookups up to date with additions and name changes 

### PR DESCRIPTION
**Pull request by jonsequitur**
This will fix the following issues:

* #260
* #340
* #1035

Changes include:

- [x] Remove `Argument.AddAlias` method, since Arguments don't really support aliases
- [x] Remove `RawAliases`
- [x] Change the meaning of `Aliases` to mean the literal aliases, including prefixes. They were only ever different for `Option`. The most important significance of this change can be seen in many of the tests which is that you can no longer query for an option without the prefix and get a positive result (e.g. given a parse result containing option `-x`, `HasOption("x")` now returns `false`, `CommandResult.ValueForOption("x")` won't find anything, etc.)
- [x] Add `Option.HasAliasIgnorePrefix`
- [x] Refactor and make perf improvements.

### CODE REVIEW

File: src/System.CommandLine/Command.cs [hunk 1]
```diff
@@ -10,12 +10,29 @@
...
+        Symbol, 
+        ICommand, 
+        IEnumerable<Symbol>
     {
         private readonly SymbolSet _globalOptions = new SymbolSet();
 
-        public Command(string name, string? description = null) : base(new[] { name }, description)
+        public Command(string name, string? description = null) : base(description)
         {
+            Name = name;
```
Comment (mikem8361) [inline src/System.CommandLine/Command.cs:22]: Won't this add the alias in the Name property setter? Doesn't it make the below AddAlias redundant?

File: src/System.CommandLine/Command.cs [hunk 2]
```diff
@@ -10,12 +10,29 @@
...
+        }
+
+        public override string Name
+        {
+            // FIX: (Name) 
+            get => base.Name;
+            set
+            {
+                base.Name = value;
+                AddAlias(Name);
```
Comment (mikem8361) [inline src/System.CommandLine/Command.cs:34]: Does the original base.Name need to be removed from the aliases? Like you do in the Options?

Comment (mikem8361) [reply src/System.CommandLine/Command.cs:34]: Never mind. It looks like it will get removed in the base (Symbol) class's Name property.

**Review by mikem8361 — APPROVED** [review]

File: src/System.CommandLine/DictionaryExtensions.cs [hunk 3]
```diff
@@ -22,5 +22,36 @@ internal static class DictionaryExtensions
...
+            }
+        }
+
+        public static bool TryRemove<TKey, TValue>(
+            this IDictionary<TKey, TValue> source,
+            TKey key)
+        {
+            if (source.ContainsKey(key))
+            {
+                source.Remove(key);
```
Comment (Keboo) [inline src/System.CommandLine/DictionaryExtensions.cs:48]: IDictionary.Remove already returns a bool. Is this method needed?

File: src/System.CommandLine/Option.cs [hunk 4]
```diff
@@ -8,18 +8,34 @@
...
+        public Option(string[] aliases, string? description = null) : base(description)
         {
+            if (aliases is null)
+            {
+                throw new ArgumentNullException(nameof(aliases));
+            }
+
+            if (!aliases.Any())
+            {
+                throw new ArgumentException("An option must have at least one alias.");
```
Comment (Keboo) [inline src/System.CommandLine/Option.cs:31]:
```suggestion
                throw new ArgumentException("An option must have at least one alias.", nameof(aliases));
```

File: src/System.CommandLine/Symbol.cs [hunk 5]
```diff
@@ -145,17 +122,38 @@ public bool HasAlias(string alias)
...
         ISymbolSet ISymbol.Children => Children;
+
+        internal Action<ISymbol>? OnNameOrAliasChanged;
+
+        [DebuggerStepThrough]
+        private protected void ThrowIfAliasIsInvalid(string alias)
+        {
+            if (string.IsNullOrWhiteSpace(alias))
+            {
+                throw new ArgumentException("An alias cannot be null, empty, or consist entirely of whitespace.");
```
Comment (Keboo) [inline src/System.CommandLine/Symbol.cs:147]:
```suggestion
                throw new ArgumentException("An alias cannot be null, empty, or consist entirely of whitespace.", nameof(alias));
```

File: src/System.CommandLine/Symbol.cs [hunk 6]
```diff
@@ -145,17 +122,38 @@ public bool HasAlias(string alias)
...
+            }
+
+            for (var i = 0; i < alias!.Length; i++)
+            {
+                if (char.IsWhiteSpace(alias[i]))
+                {
+                    throw new ArgumentException($"{GetType().Name} alias cannot contain whitespace: \"{alias}\"");
```
Comment (Keboo) [inline src/System.CommandLine/Symbol.cs:154]:
```suggestion
                    throw new ArgumentException($"{GetType().Name} alias cannot contain whitespace: \"{alias}\"", nameof(alias));
```

File: src/System.CommandLine/Symbol.cs [hunk 7]
```diff
@@ -145,17 +122,38 @@ public bool HasAlias(string alias)
...
+
+        [DebuggerStepThrough]
+        private protected void ThrowIfAliasIsInvalid(string alias)
+        {
+            if (string.IsNullOrWhiteSpace(alias))
+            {
+                throw new ArgumentException("An alias cannot be null, empty, or consist entirely of whitespace.");
+            }
+
+            for (var i = 0; i < alias!.Length; i++)
```
Comment (Keboo) [inline src/System.CommandLine/Symbol.cs:150]: Is the null forgiveness operator needed here? The parameter is declared non-nullable.

**Comment by mikem8361** [comment]
Can you let me know what build version these changes made it into?

**Pull request merged**